src/Mono.Android/Android.Runtime/JNINativeWrapper.cs

```diff
@@ -51,26 +51,34 @@ public static Delegate CreateDelegate (Delegate dlg)
 
 			ig.Emit (OpCodes.Call, wait_for_bridge_processing_method!);
 
-			var label = ig.BeginExceptionBlock ();
+			bool  filter = Debugger.IsAttached || !JNIEnv.PropagateExceptions;
+			if (!filter) {
```

Comment:
@thaystg: This presents a question: is there any way to "mimic" the "normal" .NET Unhandled Exception behaviors while still catching the exception?

Related: https://github.com/xamarin/xamarin-android/issues/4548

The problem is that when the `catch` block doesn't exist, in a managed > Java > managed callstack, if the rightmost frame throws (and no `catch` block is present around it) and execution *doesn't* return to Java, and we unwind the rightmost Java > managed frames, then JVM state may be corrupted.

In an ideal world, *every* "marshal method" would have a `try{…} catch (Exception) {…}` handler, so that when a managed method throws an exception it can be properly caught & propagated back to Java. The problem with this ideal is it means every exception is *by definition* "handled", thus "mooting" the entire concept of "*first chance* unhandled exceptions".  (There'd still *eventually* be an unhandled exception, but that would be on the *leftmost* managed frame, after the rightmost frame had been unwound.)

Is this possible?

Reply:
Are you telling me that removing the try catch when the debugger is attached is a problem because when we have a case like:
managed -> Java -> managed (which throws the unhandled exception), debugger will stop in the exception, but when we continue the execution we may get an error because JVM state may be corrupted? Which kind of error? shouldn't the process be finished in this case?
If I test #4548 I will reproduce the case that you are saying?

Reply:
> Are you telling me that removing the try catch when the debugger is attached is a problem because when we have a case like:
> managed -> Java -> managed (which throws the unhandled exception), debugger will stop in the exception, but when we continue the execution we may get an error because JVM state may be corrupted?

Yup. Because the debugger is running, there are no `try`/`catch` blocks at Java-to-managed boundaries, and because there are `try`/`catch` blocks, *if* there is an exception thrown from managed code, the JVM state *is* corrupted, and Android aborts the process.

*Desktop* repro:

```bash
$ git clone https://github.com/xamarin/java.interop.git
$ cd java.interop
$ make prepare
$ cat > d <<EOF
diff --git a/tests/Java.Interop-Tests/Java.Interop/TestType.cs b/tests/Java.Interop-Tests/Java.Interop/TestType.cs
index 4640d5f1..2cd108b1 100644
--- a/tests/Java.Interop-Tests/Java.Interop/TestType.cs
+++ b/tests/Java.Interop-Tests/Java.Interop/TestType.cs
@@ -172,7 +172,8 @@ namespace Java.InteropTests
 		static Delegate GetMethodThrowsHandler ()
 		{
 			Action<IntPtr, IntPtr> h = MethodThrowsHandler;
-			return JniEnvironment.Runtime.MarshalMemberBuilder.CreateMarshalToManagedDelegate (h);
+//			return JniEnvironment.Runtime.MarshalMemberBuilder.CreateMarshalToManagedDelegate (h);
+			return h;
 		}
 
 		static void MethodThrowsHandler (IntPtr jnienv, IntPtr n_self)

EOF
$ git apply < d
$ make
$ make run-tests TESTS=bin/TestDebug/Java.Interop-Tests.dll
```

Desired behavior: nothing bad happens. ;-)

Actual results: horrible flaming death:

```bash
$ make run-tests TESTS=bin/TestDebug/Java.Interop-Tests.dll
r=0; \
	 	msbuild /p:Configuration=Debug  build-tools/scripts/RunNUnitTests.targets /p:TestAssembly=bin/TestDebug/Java.Interop-Tests.dll || r=1; \
	exit $r;
…
  =================================================================
  	Native Crash Reporting
  =================================================================
  Got a segv while executing native code. This usually indicates
  a fatal error in the mono runtime or one of the native libraries 
  used by your application.
  =================================================================
  
  =================================================================
  	Native stacktrace:
  =================================================================
  	0x101da3779 - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : mono_dump_native_crash_info
  	0x101d3b5be - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : mono_handle_native_crash
  	0x101d9d8f8 - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : altstack_handle_and_restore
  	0x10c849e4e - /Users/jon/android-toolchain/jdk-11/lib/server/libjvm.dylib : _ZN5Chunk9next_chopEv
  	0x10cb5ccec - /Users/jon/android-toolchain/jdk-11/lib/server/libjvm.dylib : checked_jni_GetBooleanField
  	0x10c299140 - /Users/jon/Dropbox/Developer/Java.Interop/bin/TestDebug/libjava-interop.dylib : java_interop_jnienv_get_boolean_field
  	0x1650e3841 - Unknown
  
  =================================================================
  	Telemetry Dumper:
  =================================================================
EXEC : warning : SIGSEGV handler expected:libjvm.dylib+0x5f61c3  found:0x0000000000000000 [/Users/jon/Dropbox/Developer/Java.Interop/build-tools/scripts/RunNUnitTests.targets]
  Signal Handlers:
  SIGSEGV: SIG_DFL, sa_mask[0]=00000000000000000000000000000000, sa_flags=none
…
```

Again, this crash is "expected", because by *not* catching all exceptions (which is what `JniEnvironment.Runtime.MarshalMemberBuilder.CreateMarshalToManagedDelegate()` does), we're propagating a Mono exception "through" a Java-to-managed boundary, and Java Does Not Like That. (Deservedly so!)

The Good And Proper solution is to catch *all* exceptions at *all* Java-to-managed boundaries, but in doing so we "kill" the normal/expected IDE unhandled exception behavior… because now all exceptions are handled.

Is there a way to have our cake and eat it, too?

Reply:
@thaystg what is the best path forward here?

Reply:
I'm not sure how can we do this, but I would like to try to generate a try catch with some "mark" so we would know that the debugger should stop because it is not a really handled exception, it's only an exception that we catch in our forced catch.
I will come back to soon to this issue to make more tests.